### ISSUE TITLE
java.lang.AssertionError: assertion failed

### ISSUE THREAD

**Issue by legraphista** [neo4j/neo4j#8287]
I discovered this by accident while playing around with `NULL`

- Neo4j version: 3.0.6
- Operating system: Ubuntu 16.04
- API/Driver: Neo4j Browser
- **Steps to reproduce**
```cypher
UNWIND [{test: 1}] AS x WITH x
WHERE 
  NULL = x.text OR
    (
      COALESCE(NULL <> x.test, true)  AND
      NOT COALESCE(null, false)
    )
return x
```
- Expected behavior
```
╒═════════╕
│x        │
╞═════════╡
│{test: 1}│
└─────────┘
```
- Actual behavior
```
assertion failed
```
## Logs and info:

Same query works on 3.0.0

Here's the equivalent query that works around the bug:
```
UNWIND [{test: 1}] AS x WITH x
WHERE 
	NULL = x.text OR
    (
	COALESCE(NULL, 0) <> x.test  AND
	NOT COALESCE(null, false)
	)
return x
```

Here's a stack trace: 
```
2016-10-31 20:21:50.719+0000 ERROR [o.n.b.v.r.i.ErrorReporter] Client triggered an unexpected error [UnknownError]: assertion failed, reference 256327cb-7279-4784-a560-2c306d895070. assertion failed
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.SetChecker.<init>(Checker.scala:78)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.BuildUp.checkAndBuildUpCache(Checker.scala:63)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.BuildUp.contains(Checker.scala:46)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.InCheckContainer.contains(ConstantCachedIn.scala:112)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.DynamicCachedIn.isMatch(ConstantCachedIn.scala:87)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.Not.isMatch(Predicate.scala:103)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.Predicate.apply(Predicate.scala:38)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.expressions.CoalesceFunction$$anonfun$1.apply(CoalesceFunction.scala:31)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.expressions.CoalesceFunction$$anonfun$1.apply(CoalesceFunction.scala:31)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$class.find(Iterator.scala:944)
	at scala.collection.AbstractIterator.find(Iterator.scala:1336)
	at scala.collection.IterableLike$class.find(IterableLike.scala:79)
	at scala.collection.SeqViewLike$AbstractTransformed.find(SeqViewLike.scala:37)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.expressions.CoalesceFunction.apply(CoalesceFunction.scala:32)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CoercedPredicate.isMatch(Predicate.scala:316)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1$$anonfun$2.apply(Predicate.scala:81)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1$$anonfun$2.apply(Predicate.scala:81)
	at scala.util.Try$.apply(Try.scala:192)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1.apply(Predicate.scala:81)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1.apply(Predicate.scala:76)
	at org.neo4j.cypher.internal.frontend.v3_0.helpers.NonEmptyList$class.foldLeftLoop(NonEmptyList.scala:256)
	at org.neo4j.cypher.internal.frontend.v3_0.helpers.NonEmptyList$class.foldLeft(NonEmptyList.scala:183)
	at org.neo4j.cypher.internal.frontend.v3_0.helpers.Fby.foldLeft(NonEmptyList.scala:302)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate.isMatch(Predicate.scala:76)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1$$anonfun$2.apply(Predicate.scala:81)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1$$anonfun$2.apply(Predicate.scala:81)
	at scala.util.Try$.apply(Try.scala:192)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1.apply(Predicate.scala:81)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate$$anonfun$1.apply(Predicate.scala:76)
	at org.neo4j.cypher.internal.frontend.v3_0.helpers.NonEmptyList$class.foldLeftLoop(NonEmptyList.scala:256)
	at org.neo4j.cypher.internal.frontend.v3_0.helpers.NonEmptyList$class.foldLeft(NonEmptyList.scala:183)
	at org.neo4j.cypher.internal.frontend.v3_0.helpers.Fby.foldLeft(NonEmptyList.scala:302)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.CompositeBooleanPredicate.isMatch(Predicate.scala:76)
	at org.neo4j.cypher.internal.compiler.v3_0.commands.predicates.Predicate.apply(Predicate.scala:38)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.ProjectionPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(ProjectionPipe.scala:48)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.ProjectionPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(ProjectionPipe.scala:46)
	at scala.collection.immutable.Map$Map2.foreach(Map.scala:137)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.ProjectionPipe$$anonfun$internalCreateResults$1.apply(ProjectionPipe.scala:46)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.ProjectionPipe$$anonfun$internalCreateResults$1.apply(ProjectionPipe.scala:45)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:462)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:461)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:461)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$class.isEmpty(Iterator.scala:330)
	at scala.collection.AbstractIterator.isEmpty(Iterator.scala:1336)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.TopNPipe.internalCreateResults(TopPipe.scala:76)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.PipeWithSource.createResults(Pipe.scala:126)
	at org.neo4j.cypher.internal.compiler.v3_0.pipes.PipeWithSource.createResults(Pipe.scala:123)
	at org.neo4j.cypher.internal.compiler.v3_0.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:98)
	at org.neo4j.cypher.internal.compiler.v3_0.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:68)
	at org.neo4j.cypher.internal.compiler.v3_0.executionplan.InterpretedExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:143)
	at org.neo4j.cypher.internal.compiler.v3_0.executionplan.InterpretedExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:127)
	at org.neo4j.cypher.internal.compiler.v3_0.executionplan.InterpretedExecutionPlanBuilder$$anon$1.run(ExecutionPlanBuilder.scala:102)
	at org.neo4j.cypher.internal.compatibility.CompatibilityFor3_0$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor3_0.scala:274)
	at org.neo4j.cypher.internal.compatibility.CompatibilityFor3_0$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor3_0.scala:272)
	at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor3_0$.runSafely(CompatibilityFor3_0.scala:188)
	at org.neo4j.cypher.internal.compatibility.CompatibilityFor3_0$ExecutionPlanWrapper.run(CompatibilityFor3_0.scala:272)
	at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:27)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:111)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:65)
	at org.neo4j.bolt.v1.runtime.internal.CypherStatementRunner.run(CypherStatementRunner.java:73)
	at org.neo4j.bolt.v1.runtime.internal.StandardStateMachineSPI.run(StandardStateMachineSPI.java:110)
	at org.neo4j.bolt.v1.runtime.internal.SessionStateMachine$State$2.runStatement(SessionStateMachine.java:132)
	at org.neo4j.bolt.v1.runtime.internal.SessionStateMachine.run(SessionStateMachine.java:725)
	at org.neo4j.bolt.v1.runtime.internal.concurrent.SessionWorkerFacade.lambda$run$1(SessionWorkerFacade.java:69)
	at org.neo4j.bolt.v1.runtime.internal.concurrent.SessionWorker.execute(SessionWorker.java:116)
	at org.neo4j.bolt.v1.runtime.internal.concurrent.SessionWorker.run(SessionWorker.java:77)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by spacecowboy** [comment]
Thank you for the clear report. We'll look into it

**Comment by henriknyman** [comment]
This has been fixed in 3.0.7 by https://github.com/neo4j/neo4j/pull/8053
